Set -march= flag based on available features rather than the base config

leekillough opened the pull request:
Set `-march=` flag based on available features rather than forcing it to the base configuration (`rv32i`, `rv32iv`, `rv64i`, `rv64iv`). Might be the same as leaving `-march=` out entirely.

leekillough committed:
Set -march flag based on available features rather than forcing it to the base configuration (rv32i, rv32iv, rv64i, rv64iv)

leekillough committed:
Handle when _riscv_arch_test isn't defined

leekillough committed:
fix logic

angsch commented on frame/base/bli_riscv_detect_arch.h:
```diff
@@ -0,0 +1,148 @@
+#if __riscv_arch_test && ( __riscv_i && __riscv_m && __riscv_a && \
+						   __riscv_f && __riscv_d && !__riscv_e)
+
+CAT(rv, CAT(__riscv_xlen, CAT(g, CAT(RISCV_C, CAT(RISCV_P, RISCV_V)))))
```
I am not sure if this is fully correct. My understanding is that `g` subsumes all of the extensions, but covers more.

https://five-embeddev.com/riscv-isa-manual/latest/naming.html#instruction-set-extension-names:
> We have also defined an abbreviation “G” to represent the “IMAFDZicsr_Zifencei” base and extensions, as this is intended to represent our standard general-purpose ISA.

leekillough replied:
I don't know how to test for `Zicsr` or `Zfencei` (or the all-encompassing G pseudo-extension) with the C API. I can remove the G shorthand and just list the individual extensions which we can detect.

angsch commented on frame/base/bli_riscv_detect_arch.h:
```diff
@@ -0,0 +1,148 @@
+#else /* __riscv_arch_test */
+
+/* We assume I and E are exclusive when __riscv_arch_test isn't defined */
+#if __riscv_32e
```
We are in the `else` branch here, ie the compiler does **not** support the architecture extension test macro. Why does `__riscv_32e` work anyway?

leekillough replied:
> We are in the `else` branch here, ie the compiler does **not** support the architecture extension test macro. Why does `__riscv_32e` work anyway?
<html><body><!--StartFragment-->

Name | Value | When defined
-- | -- | --
__riscv_32e | 1 | E extension is available.
<!--EndFragment-->
</body>
</html>

[Source](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#preprocessor-definitions)

angsch commented on frame/base/bli_riscv_detect_arch.h:
```diff
@@ -0,0 +1,148 @@
+
+CAT(rv, CAT(__riscv_xlen, CAT(RISCV_I, CAT(RISCV_E, CAT(RISCV_M, CAT(RISCV_A,
+CAT(RISCV_F, CAT(RISCV_D, CAT(RISCV_C, CAT(RISCV_P, RISCV_V))))))))))
+
```
Have you been able to test the `else` branch? I recall haven gotten a compiler error at some point, complaining that the extensions have to be listed in a certain order. In particular, the order of `p` and `v` is not what I would have expected. I assume that `p` requires the presence of `v` - then the logical order would be `vp`, but I do not know what is correct. Note that this comment also applies to line 139.

leekillough replied:
`P` is listed before `V`, and does not imply `V`. `P` supports DSP SIMD extensions, and is unrelated to `V` vector extensions.

In Table 27.1, page 152, of the [Unprivileged ISA](https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMAFDQC/riscv-spec-20191213.pdf), it lists `P` before `V`. It also lists `__riscv_p` before `__riscv_v` in the [C API spec](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#user-content-architecture-extension-test-macro).

> Table 27.1: Standard ISA extension names. The table also defines the canonical order in which
> extension names must appear in the name string, with top-to-bottom in table indicating first-to-last
> in the name string, e.g., RV32IMACV is legal, whereas RV32IMAVC is not.

`rv64imafdpv` is legal, while `rv64imafdvp` is not.

angsch commented on frame/base/bli_riscv_detect_arch.h:
```diff
@@ -0,0 +1,148 @@
+#define RISCV_C
+#endif
+
+#if __riscv_p
```
As far as I know, `p` is not yet ratified. I am tempted to skip `p` for the time being. See the comment at the bottom about the relationship between `v` and `p`.

leekillough replied:
`P` is defined in the C API spec, which is why it's included here.

leekillough committed:
Remove g shorthand notation, always listing individual extensions.

If __riscv_arch_test is not defined, fall back on __riscv_mul,
__riscv_atomic, and __riscv_compressed to test for M, A and C
extensions.

leekillough committed:
Fix typo
Change bli_static_assert() to use STATIC_ASSERT_FAILED name in bitfield.

leekillough commented on frame/include/bli_misc_macro_defs.h:
```diff
@@ -171,6 +171,6 @@ BLIS_INLINE void bli_toggle_bool( bool* b )
 
 // Static assertion compatible with any version of C/C++
-#define bli_static_assert(cond) while(0){struct s {int bli_static_assert : !!(cond);};}
+#define bli_static_assert(cond) while(0){struct s {int STATIC_ASSERT_FAILED : !!(cond);};}
```
https://github.com/flame/blis/pull/693/files#r1147028203
